pk/ecc/ecc_sign.c

```diff
@@ -0,0 +1,350 @@
+    }
+
+    /* Verify ECC Signature */
+    if (ret == 0) {
+        ret = wc_ecc_verify_hash_ex(
+            &r, &s,       /* r/s as mp_int */
+            hash, hashSz, /* computed hash digest */
+            &verify_res,  /* verification result 1=success */
+            &ecc
+        );
+    }
+    
+    /* check verify result */
+    if (ret == 0 && verify_res == 0) {
+        ret = SIG_VERIFY_E;
+    }
+
+    mp_clear(&r);
+    mp_clear(&s);
+    wc_ecc_free(&ecc);
+
+    return ret;
+}
+#endif /* HAVE_ECC_VERIFY */
+
+#ifdef HAVE_ECC_SIGN
+/* perform signature operation against hash using private key */
+int crypto_ecc_sign(const uint8_t *key, uint32_t keySz,
+    const uint8_t *hash, uint32_t hashSz, uint8_t *sig, uint32_t* sigSz,
+    int curveSz, int curveId)
+{
+    int ret;
+    mp_int r, s;
+    ecc_key ecc;
+    WC_RNG rng;
+
+    /* validate arguments */
+    if (key == NULL || hash == NULL || sig == NULL || sigSz == NULL ||
+        curveSz == 0 || hashSz == 0 || keySz < curveSz || *sigSz < (curveSz*2))
+    {
+        return BAD_FUNC_ARG;
+    }
+
+    /* Initialize signature result */
+    memset(sig, 0, curveSz*2);
+
+    /* Setup the RNG */
+    ret = wc_InitRng(&rng);
+    if (ret < 0) {
+        return ret;
+    }
+
+    /* Setup the ECC key */
+    ret = wc_ecc_init(&ecc);
+    if (ret < 0) {
+        wc_FreeRng(&rng);
+        return ret;
+    }
+
+    /* Setup the signature r/s variables */
+    ret = mp_init(&r);
+    if (ret != MP_OKAY) {
+        wc_ecc_free(&ecc);
+        wc_FreeRng(&rng);
+        return ret;
+    }
+    ret = mp_init(&s);
+    if (ret != MP_OKAY) {
+        mp_clear(&r);
+        wc_ecc_free(&ecc);
+        wc_FreeRng(&rng);
+        return ret;
+    }
+
+    /* Import private key "k" */
+    ret = wc_ecc_import_private_key_ex(
+        key, keySz, /* private key "d" */
+        NULL, 0,    /* public (optional) */
+        &ecc,
+        curveId     /* ECC Curve Id */
+    );
+    if (ret == 0) {
+        /* Verify ECC Signature */
+        ret = wc_ecc_sign_hash_ex(
+            hash, hashSz, /* computed hash digest */
+            &rng, &ecc,   /* random and key context */
+            &r, &s        /* r/s as mp_int */
+        );
+
+        /* export r/s */
+        mp_to_unsigned_bin(&r, sig);
+        mp_to_unsigned_bin(&s, sig + curveSz);
+    }
+    
+    mp_clear(&r);
+    mp_clear(&s);
+    wc_ecc_free(&ecc);
+    wc_FreeRng(&rng);
+
+    return ret;
+}
+#endif /* HAVE_ECC_SIGN */
+#endif /* HAVE_ECC */
+
+static void print_hex(uint8_t* data, int sz)
+{
+    int i;
+    for (i = 0; i < sz; i++) {
+        printf("%02X ", data[i]);
+        if (i > 0 && ((i+1) % 16) == 0)
+            printf("\n");
+    }
+    printf("\n");
+}
+
+int main()
+{
+#if defined(HAVE_ECC) && defined(HAVE_ECC_SIGN) && defined(HAVE_ECC_VERIFY) && \
+    !defined(NO_SHA256)
+    int ret;
+    uint8_t hash[WC_SHA256_DIGEST_SIZE];
+    uint8_t sig[ECC_CURVE_SZ*2];
```

Comment:
should be initialized to zero for the call to print_hex in the event crypto_sha256 fails since we don't exit.

```
uint8_t sig[ECC_CURVE_SZ*2] = {0};
```